tests/e2e-kubernetes/testsuites/credentials.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	expectWriteToFail := func(pod *v1.Pod) {
+		seed := time.Now().UTC().UnixNano()
+		path := "/mnt/volume1/file.txt"
+		ginkgo.By(fmt.Sprintf("checking if writing to %s fails", path))
+		checkWriteToPathFails(f, pod, path, toWrite, seed)
+	}
+
+	ginkgo.Context("Driver level", func() {
+		// TODO: Add tests for current functionality.
+	})
+
+	ginkgo.Context("Pod level", func() {
+		ginkgo.Context("should use correct credentials", func() {
+			ginkgo.It("full access role", func(ctx context.Context) {
+				sa, deleteSA := createServiceAccount(ctx, f, "s3-csi-e2e-sa", annotateServiceAccountWithRole(iamRoleS3FullAccess))
+				defer deleteSA(ctx)
+
+				resource := storageframework.CreateVolumeResource(ctx, driver, l.config, pattern, t.GetTestSuiteInfo().SupportedSizeRange)
+				l.resources = append(l.resources, resource)
+
+				pod, err := createPodWithServiceAccount(ctx, f.ClientSet, f.Namespace.Name, []*v1.PersistentVolumeClaim{resource.Pvc}, sa.Name)
+				framework.ExpectNoError(err)
+				defer func() {
+					framework.ExpectNoError(e2epod.DeletePodWithWait(ctx, f.ClientSet, pod))
+				}()
+
+				path := expectWriteToSucceed(pod)
+				expectReadToSucceed(pod, path)
+			})
+
+			ginkgo.It("read-only role", func(ctx context.Context) {
+				sa, deleteSA := createServiceAccount(ctx, f, "s3-csi-e2e-sa", annotateServiceAccountWithRole(iamRoleS3ReadOnlyAccess))
+				defer deleteSA(ctx)
+
+				resource := storageframework.CreateVolumeResource(ctx, driver, l.config, pattern, t.GetTestSuiteInfo().SupportedSizeRange)
+				l.resources = append(l.resources, resource)
+
+				pod, err := createPodWithServiceAccount(ctx, f.ClientSet, f.Namespace.Name, []*v1.PersistentVolumeClaim{resource.Pvc}, sa.Name)
+				framework.ExpectNoError(err)
+				defer func() {
+					framework.ExpectNoError(e2epod.DeletePodWithWait(ctx, f.ClientSet, pod))
+				}()
+
+				expectWriteToFail(pod)
+			})
+		})
+
+		ginkgo.It("should refresh credentials after receiving new tokens", func(ctx context.Context) {
+			// TODO:
+			// 1. Trigger a manual `TokenRequest` or wait for it's own lifecylce
+			// 2. Assert new token file is written to the Pod
+		})
+
+		ginkgo.It("should use up to date role associated with service account", func(ctx context.Context) {
+			// Create a SA with full access role
+			sa, deleteSA := createServiceAccount(ctx, f, "s3-csi-e2e-sa", annotateServiceAccountWithRole(iamRoleS3FullAccess))
+			defer deleteSA(ctx)
+
+			resource := storageframework.CreateVolumeResource(ctx, driver, l.config, pattern, t.GetTestSuiteInfo().SupportedSizeRange)
+			l.resources = append(l.resources, resource)
+
+			pod, err := createPodWithServiceAccount(ctx, f.ClientSet, f.Namespace.Name, []*v1.PersistentVolumeClaim{resource.Pvc}, sa.Name)
+			framework.ExpectNoError(err)
+
+			path := expectWriteToSucceed(pod)
+			expectReadToSucceed(pod, path)
+
+			// Associate SA with read-only access role
+			saClient := f.ClientSet.CoreV1().ServiceAccounts(f.Namespace.Name)
```

Comment:
Didn't we create a utility function for this?

Reply:
There is a utility for creating SA, but this is just creates a client and updates it, we can create a util for this one as well.